GitHub GraphQL API support for github backend - Initial feature

erezrokah opened the pull request:
**Summary**

Fixes #2431.

I'm hoping to reduce number of requests and data traffic by using the [GitHub GraphQL API](https://developer.github.com/v4/) and by using the caching capabilities of the Apollo libraries.

**Test plan**

I manually tested the functionality of the CMS. I'm not sure unit tests are relevant as it will require heavy mocking and won't add much confident.

**Details**

In order to switch to the GraphQL API just add `use_graphql: true` under `backend` in your `config.yml`.

This is mostly a request for comments and I'm open to suggestions.
I made it so both APIs (GraphQL and REST) can co-exist and anyway we can't really replace the REST API completely as some features are still missing (mostly mutations, but also you can only query `refs` from the `heads`, `guest`, `tags`, `notes` namespaces, according to GithHub support).

I tried to consolidate multiple requests into one GraphQL query where applicable and to revert to the old API on failures where relevant.

This is my first attempt at a non tutorial/example GraphQL API so any feedback is welcome. I tried my best to apply best practices (e.g. updating the store on mutations, using fragments and union types)

I'll add comments to the PR on relevant parts of the code, but the main concept was to extract parts of `API.js` that can be overridden into separate functions and overrided them in `GraphQLAPI.js`.

**Tip**
You can use https://developer.github.com/v4/explorer/ to test queries. I recommend using a test repository as the queries/mutations run on real data

**P.S.**
You'll see some changes to `package.json` files that seem unrelevant. The reason is that I branched out from this [PR](https://github.com/netlify/netlify-cms/pull/2435)

erezrokah commented on .prettierignore:
```diff
@@ -1,4 +1,5 @@
-.cache/
\ No newline at end of file
+.cache/
+packages/netlify-cms-backend-github/src/fragmentTypes.js
```
`fragmentTypes.js` is auto generated so it should not be formatted

erquhart replied:
If it's auto generated we should pull it out of `src`.

erezrokah replied:
Do you mean not to commit it at all and generate it as part of the bootstrap, or just put it in a designated directory?
I think it's useful to have it versioned and consistent between developers, thus I committed it.

erezrokah commented on packages/netlify-cms-backend-github/package.json:
```diff
@@ -17,10 +17,17 @@
     "build": "cross-env NODE_ENV=production webpack",
-    "build:esm": "cross-env NODE_ENV=esm babel src --out-dir dist/esm --ignore **/__tests__ --root-mode upward"
+    "build:esm": "cross-env NODE_ENV=esm babel src --out-dir dist/esm --ignore **/__tests__ --root-mode upward",
+    "createFragmentTypes": "node scripts/createFragmentTypes.js"
```
script to generate `fragmentTypes.js`. Apollo will use `fragmentTypes.js` to better support [union types](https://www.apollographql.com/docs/react/advanced/fragments/#fragments-on-unions-and-interfaces)

erezrokah commented on packages/netlify-cms-backend-github/scripts/createFragmentTypes.js:
```diff
@@ -0,0 +1,48 @@
+  throw new Error('Missing environment variable GITHUB_API_TOKEN');
+}
+
+fetch(`${API_HOST}/graphql`, {
```
see https://www.apollographql.com/docs/react/advanced/fragments/#fragments-on-unions-and-interfaces

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -148,6 +148,14 @@ export default class API {
     });
   }
 
+  retrieveMetadataFile(key) {
```
I originally planned to replace this with a query, but the GraphQL API doesn't return anything for `refs/meta` at the moment. I'm discussing it with GitHub support

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -158,54 +166,56 @@ export default class API {
+    });
+  }
+
+  retrieveContent(path, branch) {
```
Overridden in `GraphQLAPI`

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -158,54 +166,56 @@ export default class API {
     }
   }
 
+  retrieveBlob(sha) {
```
Overridden in `GraphQLAPI`

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -345,33 +355,47 @@ export default class API {
 
-  deleteFile(path, message, options = {}) {
-    const branch = options.branch || this.branch;
+  getFileSha(path, branch) {
```
Overridden in `GraphQLAPI` without reading the entire file tree

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -345,33 +355,47 @@ export default class API {
     });
   }
 
+  createBranchAndPullRequest(branchName, sha, commitMessage) {
```
Overridden in `GraphQLAPI` and replaced by one mutation requests instead of two REST API requests

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -466,8 +491,10 @@ export default class API {
        */
-      const baseBranch = await this.getBranch();
-      const commits = await this.getPullRequestCommits(prNumber, head);
+      const [baseBranch, commits] = await Promise.all([
```
`getBranch` and `getPullRequestCommits` can run run in parallel

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -692,7 +719,7 @@ export default class API {
       method: 'PATCH',
       body: JSON.stringify({
-        state: closed,
+        state: 'closed',
```
Not sure how this worked, but `closed` existing on `window` thus it didn't fail:
![image](https://user-images.githubusercontent.com/26760571/61468645-6bc07e80-a986-11e9-95f2-fbdff3627373.png)

But actually send `state: false` and the REST API ate it and closed the pr

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+    const httpLink = createHttpLink({ uri: `${this.api_root}/graphql` });
+    return new ApolloClient({
+      link: authLink.concat(httpLink),
+      cache: new InMemoryCache({ fragmentMatcher }),
```
caching is disabled by default

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+  }
+
+  reset() {
+    return this.client.resetStore();
```
cleans cache (called when user logs out)

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+    const { data } = await this.query({
+      query: queries.repository,
+      variables: { owner, name },
+      fetchPolicy: CACHE_FIRST, // repository id doesn't change
```
Enables caching for this specific query

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+    } else if (!is_binary) {
+      return text;
+    } else {
+      return super.retrieveContent(path, branch);
```
The GraphQL docs are not really helpful on how to deal with binary data. They just mention that in case of binary data `text` is null and `is_binary` is true, so I decided to revert to the REST API in that case

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+    } else if (!is_binary) {
+      return text;
+    } else {
+      return super.retrieveBlob(sha);
```
See comment on `retrieveContent` https://github.com/netlify/netlify-cms/pull/2456/files#r304971340

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+    if (data.repository.object) {
+      const files = data.repository.object.entries.map(e => ({
+        ...e,
+        path: `${path}/${e.name}`,
```
Couldn't find a better way to build the `path` and `download_url` fields

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+    }
+  }
+
+  async listUnpublishedBranches() {
```
This single query replaces a much more complicated one in the base class (that has multiple requests)

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+    }
+  }
+
+  async readUnpublishedBranchFile(contentKey) {
```
A single query instead of multiple requests in the REST version.
The REST version will also read the entire file from the base branch just to check if it exists.

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+  }
+
+  async patchRef(type, name, sha, opts = {}) {
+    if (type !== 'heads') {
```
it seems that only `heads` is supported at the moment

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+      variables: {
+        deleteRefInput: { refId: branch.id },
+      },
+      update: store => store.data.delete(defaultDataIdFromObject(branch)),
```
remove from cache

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+    return data.deleteRef;
+  }
+
+  getPullRequestAndBranchQuery(branch, number) {
```
A single query instead of two in the base class

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+          closePullRequestInput: { pullRequestId: pullRequest.id },
+        },
+        update: store => {
+          store.data.delete(defaultDataIdFromObject(branch));
```
delete from cache

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+        metadata.pr.number,
+      );
+
+      const { data } = await this.mutate({
```
On mutation instead of two requests in the base class (close pr, then delete branch)

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+    }
+  }
+
+  async createBranchAndPullRequest(branchName, sha, title) {
```
One mutation instead of two requests in the base class (create branch, open pr)

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+    return { ...pullRequest, head: { sha: pullRequest.headRefOid } };
+  }
+
+  async getTree(sha) {
```
We can actually do a deep nested query to get the entire repo structure in one query, but since there is already some code in the base class to handle the recursion, it will require a more significant change.

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+  commit: '160000',
+};
+
+const fragmentMatcher = new IntrospectionFragmentMatcher({
```
see https://www.apollographql.com/docs/react/advanced/fragments/#fragments-on-unions-and-interfaces

erezrokah commented on packages/netlify-cms-backend-github/src/implementation.js:
```diff
@@ -100,6 +103,9 @@ export default class GitHub {
   logout() {
     this.token = null;
+    if (typeof this.api.reset === 'function') {
+      return this.api.reset();
```
will clear the GraphQL local cache

erezrokah commented on packages/netlify-cms-backend-github/src/queries.js:
```diff
@@ -0,0 +1,241 @@
+  ${fragments.blobWithText}
+`;
+
+export const unpublishedBranchFile = gql`
```
I find this very cool. In one query you can get the file from the `head` branch and check if the file exists on the `base` branch

erquhart replied:
Wow that's crazy

erezrokah commented on packages/netlify-cms-backend-github/src/GraphQLAPI.js:
```diff
@@ -0,0 +1,459 @@
+          repositoryId: id,
+        },
+      },
+      update: (store, { data: mutationResult }) => {
```
Updates the cache so next calls to `branch` and `pullRequestAndBranch` will use the cache

erquhart commented:
@erezrokah!! This is awesome!!

This is a huge PR, need to spend some time reviewing, but I've been wanting to see us migrate to GraphQL for some time now - really amazing work here!

erquhart commented:
I'll also note that another large PR targeting the GitHub backend has been under review for a while and we were just about to get it merged up: https://github.com/netlify/netlify-cms/pull/2430

I tested rebasing your work and it only resulted in four conflicts in a single file. If you wouldn't mind rebasing once that merge occurs (likely today), I'll make sure no further conflicts are introduced before this PR gets in (barring any major change requests).

erezrokah commented:
Hey @erquhart I appreciate the quick response, and I know it will take some time to review.
I'll rebase and fix the conflicts once #2430 is merged and will make any necessary adjustments to the GraphQL part to support it.

erezrokah commented:
Note: I had to do a force push, since I rebased with `master` after https://github.com/netlify/netlify-cms/pull/2435 was merged (since bb43c6a26a487a58d05a6a0f27f8efc942145e13 was merged and 30d539807b016f8889e3ef5b5cbdca5a4cfce925 was not merged)

erquhart commented:
Need to figure out why that failed build won't run, retry isn't working either. Working through my notifications queue, I'll circle back.

erezrokah commented:
Thanks @erquhart. 
Since I'll need to rebase with `master` (and make some modifications to support the fork workflow) once #2430 is merged, I think figuring the build error can wait :)

erquhart commented:
Yeah, seems to have sorted anyway. I'll get the forking workflow PR merged and released tomorrow.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

erquhart commented:
Update: fork workflow is merged 🎉

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

erezrokah commented:
Great :) Hopefully I'll manage to do the rebase and implement the forking workflow by early next week.

erezrokah commented:
**Still not ready, for review**. I just rebased with master and added github backend e2e tests.

I think there are some bugs with the fork workflow pr (I'll post comments after I investigate more).

The github backend e2e tests are currently skipped as they will fail with some environment settings.

The basic flow is:
1. Create a test repo based on a reference repo (I'm using https://github.com/erezrokah/gatsby-netlify-cms-aws as  a reference).
2. Clone the reference repo and push it to the test repo.
3. Use test repo as the backend.
4. Delete repo when done

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -20,15 +20,19 @@ export default class API {
     this.repoURL = `/repos/${this.repo}`;
-    this.originRepoURL = this.originRepo && `/repos/${this.originRepo}`;
+    // when not in 'useForkWorkflow' mode originRepoURL === repoURL
+    this.originRepoURL = `/repos/${this.originRepo}`;
```
this saves most of the `useForkWorkflow` conditions spread throughout the file

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -115,18 +119,12 @@ export default class API {
     return `${this.repo}/${collectionName}/${slug}`;
   }
 
-  generateBranchNameFromCollectionAndSlug(collectionName, slug) {
```
Unused method, and in any case `generateContentKey` doesn't return a promise so this method will fail.

Benaiah replied:
Removed in #2523: https://github.com/netlify/netlify-cms/pull/2523/commits/3daa87dc954e95e3a784b52b64bac26f0a8c8f6b

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -115,18 +119,12 @@ export default class API {
 
   branchNameFromRef(ref) {
-    return ref.substring('refs/heads/'.length - 1);
+    return ref.substring('/refs/heads/'.length - 1);
```
Looks like a bug introduced in https://github.com/netlify/netlify-cms/pull/2430.
Without the forward slash after creating an entry (and refreshing the page), nothing will appear under `workflow` (the branch and the pr are created, just not displayed)

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -115,18 +119,12 @@ export default class API {
 
   branchNameFromRef(ref) {
-    return ref.substring('refs/heads/'.length - 1);
+    return ref.substring('/refs/heads/'.length - 1);
```
Looks like a bug introduced in https://github.com/netlify/netlify-cms/pull/2430.
Without the forward slash after creating an entry (and refreshing the page), nothing will appear under `workflow` (the branch and the pr are created, just not displayed)

Benaiah replied:
Fixed in #2523: https://github.com/netlify/netlify-cms/pull/2523/commits/869830993bb9dd188213ba57682b017bcde1e6c3#diff-b4ba42c688a23ed9cec5139daec2c157R146-R148

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -350,7 +356,7 @@ export default class API {
   branchHasPR = async ({ branchName, ...rest }) => {
     const prs = await this.getPRsForBranchName({ branchName, ...rest });
-    return prs.some(pr => this.branchNameFromRef(pr.head.ref) === branchName);
+    return prs.some(pr => pr.head.ref === branchName);
```
BUG: Looks like a bug introduced in https://github.com/netlify/netlify-cms/pull/2430

`pr.head.ref` is not the full ref - it is the branch name, so calling `branchNameFromRef` breaks this method

Benaiah replied:
Fixed in #2523: https://github.com/netlify/netlify-cms/pull/2523/commits/9a63d11a9be15d0cd8c5a71141774b2175d9b284#diff-b4ba42c688a23ed9cec5139daec2c157R735-R738

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -710,16 +735,14 @@ export default class API {
    * Get a pull request by PR number.
    */
   getPullRequest(prNumber) {
```
This method was not used anywhere, and in any case I think it should use `originRepoURL`. 
I fixed the API call to use `originRepoURL` and used it in `updateUnpublishedEntryStatus` and `getUpdatedForkWorkflowMetadata`

Benaiah replied:
Fixed in #2523: https://github.com/netlify/netlify-cms/commit/9a63d11a9be15d0cd8c5a71141774b2175d9b284#diff-b4ba42c688a23ed9cec5139daec2c157R735-R738

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -809,8 +832,8 @@ export default class API {
-  publishUnpublishedEntry(collection, slug) {
-    const contentKey = this.generateContentKey(collection.get('name'), slug);
+  publishUnpublishedEntry(collectionName, slug) {
+    const contentKey = this.generateContentKey(collectionName, slug);
```
BUG: Looks like a bug introduced in https://github.com/netlify/netlify-cms/pull/2430

The first argument is actually a string and publishing an entry fails without this fix

Benaiah replied:
Fixed in #2523: https://github.com/netlify/netlify-cms/pull/2523/commits/9a63d11a9be15d0cd8c5a71141774b2175d9b284#diff-b4ba42c688a23ed9cec5139daec2c157L835-R836

erezrokah commented on packages/netlify-cms-core/src/actions/mediaLibrary.js:
```diff
@@ -253,6 +253,7 @@ export function loadMediaDisplayURL(file) {
     if (typeof url === 'string' || typeof displayURL === 'string') {
       dispatch(mediaDisplayURLRequest(id));
       dispatch(mediaDisplayURLSuccess(id, displayURL));
+      return;
```
BUG: Looks like a bug introduced in https://github.com/netlify/netlify-cms/pull/2430

This `return` was removed as a part of the pr.

Without it you can see the following errors in the `console` when trying to load the media libary:

`Error: getMediaDisplayURL is not implemented by the current backend, but the backend returned a displayURL which was not a string!`

As a result no images are displayed.

Benaiah replied:
Fixed in #2523: https://github.com/netlify/netlify-cms/pull/2523/commits/d42d27480d4ba2db923a4101a4f0a86eb9bb2f49

erezrokah commented:
@erquhart I added some comments prefixed with `BUG: ` in places I think the fork workflow pr introduced bugs.

They only thing left for me now is to re-test the GraphQL API (with and without fork workflow)

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -500,47 +505,65 @@ export default class API {
     const contentKey = this.generateContentKey(options.collectionName, entry.slug);
     const branchName = this.generateBranchName(contentKey);
     const unpublished = options.unpublished || false;
-    const branchData = await this.getBranch();
```
BUG: Looks like a bug introduced in https://github.com/netlify/netlify-cms/pull/2430

This breaks updating an existing entry since instead of rebasing the pull request branch it will try to rebase master

Benaiah replied:
Fixed in #2523: https://github.com/netlify/netlify-cms/pull/2523/commits/aa664adf8794a3050af592c952388b71b9ee25e4

erezrokah commented on packages/netlify-cms-backend-github/src/implementation.js:
```diff
@@ -86,7 +88,7 @@ export default class GitHub {
     var repoExists = false;
     while (!repoExists) {
       repoExists = await fetch(`${this.api_root}/repos/${repo}`, {
-        headers: { Authorization: `Bearer ${token}` },
```
BUG: Looks like a bug introduced in https://github.com/netlify/netlify-cms/pull/2430

The correct way to call the API is using `token ${token}` (this is done properly in `API.js`)

Benaiah replied:
Fixed in #2523: https://github.com/netlify/netlify-cms/pull/2523/commits/1bedf90a267a45e07d1e343f550df73846e46eb7

erezrokah commented:
@erquhart **I think the pr is read for review**. Since I had to test many scenarios I added more e2e tests - also to the fork workflow. The fork workflow confused me in the beginning as I expected to be able to publish an entry created from a fork directly in the CMS UI. After reading the docs I realized one should merge it directly from the GitHub UI.

Anyway the tests are skipped at the moment as they require the following `.env` file:
```
GITHUB_REPO_OWNER=owner # I use erezrokah
GITHUB_REPO_NAME=name # I use gatsby-netlify-cms-aws
GITHUB_REPO_TOKEN=***************** # with permissions to create, push and delete repos.
GITHUB_FORK_WORKFLOW_OWNER=fork-owner # I use erezrokah-test
GITHUB_FORK_WORKFLOW_TOKEN=***************** # with permissions to create, push and delete repos.
```

A video of the tests running:
[![Github Backend Editorial Workflow Test Netlify CMS](https://img.youtube.com/vi/R0fcHdJdI7c/0.jpg)](https://www.youtube.com/watch?v=R0fcHdJdI7c)

erquhart commented:
@erezrokah sorry for the delay here, just getting back from some time off - going to wait until fixes are in from @Benaiah for the fork workflow (thanks so much for pointing those out!) and then we can move forward with this.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

erezrokah commented:
@erquhart Thanks for the response. Once @Benaiah fixes are merged I'll do another rebase and make sure everything works using the tests I've added

Benaiah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -500,47 +505,65 @@ export default class API {
-      const pr = this.useForkWorkflow
-        ? undefined
-        : await this.createPR(options.commitMessage, branchName);
+      const pr = await this.createBranchAndPullRequest(
```
PRs should not automatically be created for Open Authoring drafts, since they shouldn't create a PR until they are moved to "Ready to Review".

erezrokah replied:
The function name is misleading as there is a condition inside it to not create the pr when using Open Authoring. I'll try to think of a better way to do it while still being able to override it.

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -113,13 +117,10 @@ export default class API {
   }
 
   async requestAllPages(url, options = {}) {
```
This method didn't work for private repos as it didn't pass in the authorization header.
Also it didn't work at all under the cypress electron browser since using async iterators requires some babel config changes:

https://github.com/meteor/meteor/issues/9897
https://github.com/babel/babel/issues/7467

A `TypeError("Object is not async iterable")` was throw and `Symbol.asyncIterator` is undefined when using the cypress electron browser.
I'm assuming the same will happen with any older browser.

I re-wrote it no to use async iterators and passed in the headers to support private repos.

erezrokah commented on packages/netlify-cms-backend-github/src/API.js:
```diff
@@ -610,6 +633,9 @@ export default class API {
         return this.rebasePullRequest(pr.number, branchName, contentKey, metadata, commit);
+      } else if (this.useOpenAuthoring) {
+        // if a PR hasn't been created yet for the forked repo, just patch the branch
+        await this.patchBranch(branchName, commit.sha, { force: true });
```
Fixes #2593

erezrokah commented on packages/netlify-cms-lib-util/src/backendUtil.js:
```diff
@@ -56,22 +56,24 @@ export const parseLinkHeader = flow([
 ]);
 
-export const getPaginatedRequestIterator = (url, options = {}, linkHeaderRelName = 'next') => {
+export const getAllResponses = async (url, options = {}, linkHeaderRelName = 'next') => {
```
See https://github.com/netlify/netlify-cms/pull/2456/files#r318008479

erezrokah commented:
@erquhart I did another re-base, and fixed some more bugs (see my new comments).

I also found a few minor UI bugs but I'll open new PRs for them.
I'm still working on making the tests run smoothly (the multiple entries tests occasional fail for both REST and GraphQL).

erezrokah commented:
While not an accurate test on performance here are the tests run times:

| Spec                    | REST                  | GraphQL               |
|-------------------------|-----------------------|-----------------------|
| Open Authoring Disabled | 7 minutes, 1 second   | 6 minutes, 27 seconds |
| Open Authoring Enabled  | 6 minutes, 52 seconds | 6 minutes, 24 seconds |

erquhart commented on cypress/plugins/github.js:
```diff
@@ -0,0 +1,193 @@
+const Octokit = require('@octokit/rest');
```
These tests are fantastic. I'd rather see them run against a mock server implementation than GitHub's API, maybe we can swap that in while still using the existing tests? I remember reading about Apollo helping with testability but haven't looked into it much.

erezrokah replied:
Let me search for a bit, maybe we can set up a local GraphQL server or something

erquhart replied:
If it's disabled I'd be cool with getting this merged and revisiting later.

erquhart commented:
There are other PR's fixing bugs that I believe are also fixed in this PR. None of my requests are critical, so I'm probably going to merge this as is and release into beta.

erezrokah commented:
@erquhart Thank you for merging :)
I'll go over the bugs today.

I've also made some progress with the e2e tests. 
Since mocking just the GraphQL server won't be enough as we're still using some of the REST API,
I looked into a different approach of recording real traffic data and replaying it during the tests.

I've looked into these projects:
http://www.mock-server.com/
https://github.com/Netflix/pollyjs

I was able to run the tests successfully while recording traffic using http://www.mock-server.com/, but had trouble replaying.

I tried setting up https://github.com/Netflix/pollyjs with Cypress per https://github.com/Netflix/pollyjs/issues/56 but once you pass pollyjs the window object as a context it breaks the Cypress functionality for some reason. 

So the current approach I'm using is recording with http://www.mock-server.com/ and using Cypress internal [internal mocking support](https://docs.cypress.io/api/commands/route.html) to replay.
Since Cypress only supports mocking XMLHttpRequests a [workaround](https://github.com/cypress-io/cypress/issues/95) is required to support fetch, but I was able to both record and replay.
The next step will be to transform the data recorded by http://www.mock-server.com/ to match the format expected by Cypress for mocking.

erquhart commented:
Recording is a great approach. There's some testing in the GitLab backend using [Nock](https://github.com/nock/nock), which also supports recording.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>
